### PR TITLE
Refactor backend-modules-mail

### PR DESCRIPTION
**Pull request by pozylon**
Targets:
- [x] Move mailchimp and mandrill code into mail package
- [x] Decouple the mail package by moving crowdfunding-opinionated logic from the package to republik backend. Leads to a mail package with no interdependencies to auth and pgdb.
- [x] Split the big functions into smaller pieces, this will enable testability of the mail package
- [x] More dry
- [x] Tests

### CODE REVIEW

File: packages/mail/NewsletterSubscription.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    const interstConfiguration = this.constructor.interestConfiguration(interestId)
+    this.name = interstConfiguration.name
+    this.id = Buffer.from(userId + this.name).toString('base64')
+    this.subscribed = subscribed
+    this.isEligible = this.constructor.isEligibleForInterestId(interestId, roles)
+  }
+
+  static configure (config) {
+    this.interestConfigurationMap = config
+  }
```
Comment (patte) [inline packages/mail/NewsletterSubscription.js]: globals need to be clearly visible, this is not the case here. For a developer who needs to change something about the subscriptions, who doesn't have the context of this PR, would have a hard time figure out that this is a class global that needs to be set first.

Comment (patte) [reply packages/mail/NewsletterSubscription.js]: the more I think about it, a class is the wrong approach here. You would need to check in all methods if configure was called before (precondition-check), I would instead turn all class functions, into normal functions that take a config and are stateless.

File: packages/mail/lib/configure.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+const NewsletterSubscription = require('../NewsletterSubscription')
+
+module.exports = (interestConfiguration) => {
+  NewsletterSubscription.configure(interestConfiguration)
+}
```
Comment (patte) [inline packages/mail/lib/configure.js]: this is not needed if you use stateless functions

File: packages/mail/lib/sendMail.js [hunk 3]
```diff
@@ -34,18 +34,10 @@ module.exports = (mail) => {
...
-    headers: {
-      'Content-Type': 'application/json'
-    },
-    body: JSON.stringify({
-      key: process.env.MANDRILL_API_KEY,
-      message: mail
-    })
-  })
+  const mandrill = new MandrillInterface({ logger })
+  return mandrill.send(mail)
```
Comment (patte) [inline packages/mail/lib/sendMail.js]: An exported, standalone function uses an instantiated object of a file nearby for one command, I don't get it.

File: packages/mail/lib/unsubscribeFromAllNewsletters.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  if (!email) {
+    throw new EmailRequiredMailError()
+  }
+
+  const mailchimp = new MailchimpInterface({ logger })
+  return mailchimp.updateMember(email, {
+    email_address: email,
+    status: 'unsubscribed'
+  })
+}
```
Comment (patte) [inline packages/mail/lib/unsubscribeFromAllNewsletters.js]: the name of this file is misleading. I would expect it to unsubscribe an email from all interests, instead it unsubscribes an email from mailchimp completely. change to `unsubscribeEmail`

File: packages/mail/lib/updateNewsletterSubscription.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  const body = {
+    interests: {
+      [interestId]: !!subscribed
+    }
+  }
+
+  // If a user subscribes to a newsletter but their status is not subscribed,
+  // we need to set their status to 'pending' which triggers a new confirmation email
+  // from mailchimp to re-subscribe.
+  if (subscribed && status !== 'subscribed') {
```
Comment (patte) [inline packages/mail/lib/updateNewsletterSubscription.js]: note to myself: style the mailchimp mail

**Comment by patte** [comment]
despite my scepticism about the design with classes, I wanted to add that I'm very happy that you made the work of cleaning this up.

**Comment by pozylon** [comment]
Hehe i see classes generally don't make you smile ;) Of course, I'd love to discuss my design decisions this afternoon.

- The classes in this package are only internal (not exposed through lib) and are there to structure the code in a more elegant way. That way they will never get subclasses or any crazy shit like that ;) We could of course use functions and functional factories instead.

- I always think "in layers of complexity" of a code module. Adopted to the structure we find here it means that all public (API) functions need to be in lib where as all private functions need to stay out of the lib folder. That's why there is a "configure.js" which just fast-forwards to the internal structure. The same is true for logging. At a certain "depth" I think it's better not to assume that something should be logged at all, and if, how. That way you can prefix log output (log composition) or make it dependent of log level settings on the base of a public api. It encourages the use of logging on a deep level for "debug" purposes.

- Yes we should add an exception when configuration is missing

Alternative designs I could imagine:

No classes:
Use pure js objects, function factories and plain functions to structure the code rather than classes. I think it will lead to more lines of code and more bloated syntax but with very little benefits (maybe stronger "disencouragement" of inheritance), the package will still be stateful.

Functional (no globals):
If we drop global configuration and go fully stateless, it means that we have to submit the configuration data with every call to the public API. To still have DRY code in republik-backend you would have to have a higher-order function in there which injects the configuration data on every possible call. Doing that bloats the codebase of republik-backend.

**Comment by pozylon** [comment]
After some minutes of thinking I want to drop global configuration and introduce a createMail function, using globals in js is bad design and there are elegant solutions.

**Comment by pozylon** [comment]
The whole mail package is now stateless

File: packages/mail/integration.test.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
+require('dotenv').config({ path: '../../.test.env' })
```
Comment (patte) [inline packages/mail/integration.test.js]: IMHO we should move this file to republik-backend to completly keep this "republik-free".

**Pull request merged**